lib/core/src/control/io/mouse.rs

```diff
@@ -1,407 +1,144 @@
-use crate::system::web::dyn_into;
-use crate::system::web::Result;
-use crate::system::web::Error;
-use crate::system::web::ignore_context_menu;
+use crate::control::callback::*;
+use crate::display::scene::Shape;
 
-use wasm_bindgen::prelude::*;
+use enso_frp::EventEmitterPoly;
+use enso_frp::Position;
+use std::cell::RefCell;
+use std::rc::Rc;
 use wasm_bindgen::JsCast;
-use web_sys::MouseEvent;
-use web_sys::WheelEvent;
+use wasm_bindgen::JsValue;
+use wasm_bindgen::prelude::Closure;
 use web_sys::EventTarget;
-use web_sys::AddEventListenerOptions;
-use js_sys::Function;
-use nalgebra::Vector2;
-use std::rc::Rc;
-use std::cell::RefCell;
-
-
-
-// =====================
-// === EventListener ===
-// =====================
-
-/// This struct keeps the register of the event listener and unregisters it when it's dropped.
-#[derive(Debug)]
-pub struct EventListener<T:?Sized> {
-    target   : EventTarget,
-    name     : String,
-    callback : Closure<T>
-}
-
-impl<T:?Sized> EventListener<T> {
-    fn new<S:Str>(target:EventTarget, name:S, callback:Closure<T>) -> Self {
-        let name = name.as_ref().to_string();
-        Self { target,name,callback }
-    }
-}
-
-impl<T:?Sized> Drop for EventListener<T> {
-    fn drop(&mut self) {
-        let callback: &Function = self.callback.as_ref().unchecked_ref();
-        remove_event_listener_with_callback(&self.target,&self.name,&callback).ok();
-    }
-}
-
-
-
-// =============================
-// === Mouse Event Listeners ===
-// =============================
-
-/// EventListener for Mouse events.
-pub type MouseEventListener = EventListener<dyn FnMut(MouseEvent)>;
-
-/// EventListener for Wheel events.
-pub type WheelEventListener = EventListener<dyn FnMut(WheelEvent)>;
-
-
-
-// ===================
-// === MouseButton ===
-// ===================
-
-// FIXME: this does not handle all buttons (js defines 5 buttons) and assumes mouses for
-// FIXME: right hand people.
-/// An enumeration representing the mouse buttons.
-#[derive(Clone,Copy,Debug)]
-#[allow(missing_docs)]
-pub enum MouseButton {LEFT,MIDDLE,RIGHT,UNKNOWN}
 
 
 
 // =======================
-// === MouseClickEvent ===
+// === EventDispatcher ===
 // =======================
 
-/// Mouse click callback used by `MouseManager`.
-pub trait MouseClickCallback = FnMut(MouseClickEvent) + 'static;
+// TODO: Consider merging this implementation with crate::control::callback::* ones.
 
-// FIXME: "click" means mouse down and then up. This is misleading.
-/// A struct storing information about mouse down and mouse up events.
-#[derive(Clone,Copy,Debug)]
-pub struct MouseClickEvent {
-    /// The position where the MouseClickEvent occurred.
-    pub position : Vector2<f32>,
-
-    /// The button which triggered the event.
-    pub button : MouseButton
+/// Shared event dispatcher.
+#[derive(Debug,Derivative)]
+#[derivative(Clone(bound=""))]
+#[derivative(Default(bound=""))]
+pub struct EventDispatcher<T> {
+    rc: Rc<RefCell<XCallbackRegistry1<T>>>
```

Comment:
When creating `mouse2.rs` I created the `XCallbackRegistry1` because `CallbackRegistry1` had too many cosntraints. Can we fix this in this pr as well pls ?